enclave/core/globals.c

```diff
@@ -8,6 +8,119 @@
 
 extern volatile const oe_sgx_enclave_properties_t oe_enclave_properties_sgx;
 
+#if defined(__linux__)
+/**
+ *****************************************************************************
+ * Global variables and RIP-relative addressing
+
+ * In linux, by default all symbols are exported. The compiler makes a
+ * distinction between the following 3 kinds of global variables when
+ * generating position-independent-code (via -fPIC or -fpic):
+ *     1. static global:
+ *           static int g1;
+ *           int get_g1() { return g1; }
+ *
+ *        The compiler always generated RIP relative addressing for static
+ *        global variables since it knows that these variables cannot be
+ *        addressed by other compilation units:
+ *
+ *           get_g1:
+ *              movl g1(%rip), %eax
+ *              ret
+ *
+ *     2. Local global:
+ *           int g2;
+ *           int get_g2() { return g2; }
+ *
+ *        Local globals are defined in the current compilation unit.
+ *        However another compilation unit can refer to this global via an
+ *        external declaration. Therefore the compiler adds an entry for
+ *        the variable in the Global Offset Table (GOT) and uses double
+ *        indirection to access the variable:
+ *
+ *           get_g2:
+ *              movq g2@GOTPCREL(%rip), %rax
+ *              movl (%rax), %eax
+ *              ret
+ *
+ *        Even though the compiler knows that the current compilation unit
+ *        defines the variable, another definition of the variable could
+ *        already have been loaded into memory at runtime (in another
+ *        shared-library that has a variable of the same name), hence
+ *        the compiler conservatively uses RIP relative addressing even
+ *        in the current compilation unit.
+ *
+ *     3. extern global:
+ *           extern int g3;
+ *           int get_g3() { return g3; }
+ *
+ *        These are global variables defined in another compilation unit and
+ *        the addresses of these variable can only be known at run time.
+ *        Therefore the compiler always generates an entry in the GOT for
+ *        these variables.
+ *
+ *           get_g3:
+ *              movq g3@GOTPCREL(%rip), %rax
+ *              movl (%rax), %eax
+ *              ret*
+ *
+ * Note, if the source code is compiled using position independent executable
+ * flags (-fPIE or -fpie) instead of position independent code flags
+ * (-fPIC or -fpic), then the variables defined in the compilation unit are
+ * to be given preference over definitions of the same variables in shared
+ * libraries. Hence when compiled using -fPIE/-fpie, local globals
+ * (type 2 above) are also guaranteed to be generated using RIP relative
+ * addressing just like static globals.
+ *
+ *****************************************************************************
+ * Calculating enclave base address securely
```

Comment:
address -> addresses (or change to Calculating an enclave base address securely).

Reply:
Changed to:
Calculating the enclave base address securely